File: epi_judge_java/epi/PascalTriangle.java
package epi;

import epi.test_framework.EpiTest;
import epi.test_framework.GenericTest;

import java.util.ArrayList;
import java.util.List;

public class PascalTriangle {

    @EpiTest(testDataFile = "pascal_triangle.tsv")
    public static List<List<Integer>> generatePascalTriangle(int numRows) {
        List<List<Integer>> result = new ArrayList<>();
        if (numRows == 0) return result;

        result.add(List.of(1));

        for (int i = 1; i < numRows; i++) {
            List<Integer> prev = result.get(result.size() - 1);
            List<Integer> level = new ArrayList<>();
            level.add(1);
            for (int j = 0; j < prev.size() - 1; j++) {
                level.add(prev.get(j) + prev.get(j + 1));
            }

            level.add(1);
            result.add(level);
        }

        return result;
    }

    public static void main(String[] args) {
        System.exit(
                GenericTest
                        .runFromAnnotations(args, "PascalTriangle.java",
                                new Object() {
                                }.getClass().getEnclosingClass())
                        .ordinal());
    }
}

File: epi_judge_cpp/pascal_triangle.cc
#include <vector>

#include "test_framework/generic_test.h"
using std::vector;

vector<vector<int> > GeneratePascalTriangle(int num_rows) {
    vector<vector<int>> ans;
    if (num_rows == 0) return ans;
    ans.push_back({1});
    for (int level = 2; level <= num_rows; level++) {
        vector<int> next;
        next.push_back(1);
        const auto &prev = ans[level - 2];
        for (int i = 0; i < prev.size() - 1; i++) {
            next.push_back(prev[i] + prev[i + 1]);

        }
        next.push_back(1);
        ans.push_back(next);
    }
    return ans;
}

int main(int argc, char *argv[]) {
    std::vector<std::string> args{argv + 1, argv + argc};
    std::vector<std::string> param_names{"num_rows"};
    return GenericTestMain(args, "pascal_triangle.cc", "pascal_triangle.tsv",
                           &GeneratePascalTriangle, DefaultComparator{},
                           param_names);
}
